### PR TITLE
Estimate the twin transfer gas limits

### LINKED ISSUE #757: Estimate SINGLE_TWIN_RESERVED_GAS and MINIMAL_RESIDUAL_GAS
Values for `SINGLE_TWIN_RESERVED_GAS` and `MINIMAL_RESIDUAL_GAS` were estimated at the time they were introduced.
Since then, `transferTwins` changed, so these values need to be estimated.

Ideally, prepare a script that can be called on demand.

### PR DESCRIPTION
**Pull request by zajck**
Closes #757 

~~Based on #759 which should be merged before this.~~

### CODE REVIEW

File: contracts/protocol/facets/ExchangeHandlerFacet.sol [hunk 1]
```diff
@@ -828,7 +828,7 @@ contract ExchangeHandlerFacet is IBosonExchangeHandler, BuyerBase, DisputeBase {
             // Handle overall outcome: up to 200,000 gas
             // Next line would overflow if twinCount > (type(uint256).max - MINIMAL_RESIDUAL_GAS)/SINGLE_TWIN_RESERVED_GAS
             // Oveflow happens for twinCount ~ 9.6x10^71, which is impossible to achieve
-            uint256 reservedGas = twinCount * SINGLE_TWIN_RESERVED_GAS + MINIMAL_RESIDUAL_GAS;
+            uint256 reservedGas = (twinCount - 1) * SINGLE_TWIN_RESERVED_GAS + MINIMAL_RESIDUAL_GAS;
```
Comment (albertfolch-redeemeum) [inline contracts/protocol/facets/ExchangeHandlerFacet.sol:831]: why -1? @zajck

File: scripts/util/estimate-twin-transfer-limits.js [hunk 2]
```diff
@@ -0,0 +1,210 @@
+var log = console.log;
+
+// Modify the console.log to capture the contract output
+var contractLogs = [];
```
Comment (albertfolch-redeemeum) [inline scripts/util/estimate-twin-transfer-limits.js:4]:
```suggestion
let contractLogs = [];
```

File: scripts/util/estimate-twin-transfer-limits.js [hunk 3]
```diff
@@ -0,0 +1,210 @@
+var log = console.log;
```
Comment (albertfolch-redeemeum) [inline scripts/util/estimate-twin-transfer-limits.js:1]:
```suggestion
const log = console.log;
```

File: scripts/util/estimate-twin-transfer-limits.js [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+
+        // Gas left measure points
+        // 1. At the beginning of the transfer twin loop
+        if (line.includes("for (uint256 i = 0; i < twinCount")) {
+          if (previousMeasurePoint !== "invoke") {
+            throw new Error("Could not find the invoke measure point");
+          }
+          previousMeasurePoint = "loopStart";
+          loopFound = true;
+          return line + 'console.log("TwinGasEstimate1,%s,%s", i, gasleft());';
```
Comment (albertfolch-redeemeum) [inline scripts/util/estimate-twin-transfer-limits.js:57]:
```suggestion
          return line + 'console.log(`TwinGasEstimate1,${i},${gasleft()}`);';
```

(similar suggestion for the others console.log)

File: scripts/util/estimate-twin-transfer-limits.js [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+    grep: "redeemVoucher\\(\\) with bundle",
+    // grep: "uld not decrease twin supplyAvailable if supply is unlimited",
+    reporter: "min",
+  };
+  hre.config.gasReporter.enabled = false;
+
+  log("Running tests...");
+  await hre.run("test");
+
+  // log(contractLogs);
```
Comment (albertfolch-redeemeum) [inline scripts/util/estimate-twin-transfer-limits.js:141]:
```suggestion
```

File: scripts/util/estimate-twin-transfer-limits.js [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+      },
+    }),
+  };
+
+  // 2. Run tests that cover the bundles and capture the output
+  hre.config.mocha = {
+    ...hre.config.mocha,
+    testFiles: ["./test/protocol/ExchangeHandlerTest.js"],
+    grep: "redeemVoucher\\(\\) with bundle",
+    // grep: "uld not decrease twin supplyAvailable if supply is unlimited",
```
Comment (albertfolch-redeemeum) [inline scripts/util/estimate-twin-transfer-limits.js:133]:
```suggestion
```

**Review by albertfolch-redeemeum — APPROVED** [review]
left some comments but it looks good

File: scripts/util/estimate-twin-transfer-limits.js (hunk 4 above)
Comment (zajck) [reply scripts/util/estimate-twin-transfer-limits.js:57]: This part of the script preprocesses `.sol` files and solidity does not support template literals or string interpolation, so compilation fails.
We need to stick with the existing formatting.

**Review by albertfolch-redeemeum — APPROVED** [review]

**Review by mischat — APPROVED** [review]
LGTM ... great documentation there

**Pull request merged**